### DISCUSSION TITLE
Why isn't there a GenericObjectAssertions<T> type?

### DISCUSSION THREAD

**Discussion by georgiosd** [fluentassertions/fluentassertions#1709] [Q&A]
Hey guys,

I would swear this was always there but I guess I don't remember it correctly.

Right now we can do `o.Property.Should().Be(x)` but not `o.Should().Match(x => x.Property == x).And...`

I seem to remember there was a `ShouldHave` API that could do this?

Thanks

**Comment by jnyrup** [comment]
There are two overloads of `Match` in `ReferenceTypeAssertions` 

```c#
public AndConstraint<TAssertions> Match(Expression<Func<TSubject, bool>> predicate, string because = "", params object[] becauseArgs);
public AndConstraint<TAssertions> Match<T>(Expression<Func<T, bool>> predicate, string because = "", params object[] becauseArgs) where T : TSubject;
```

You probably just need to specify the type to hit the desired overload.

```c#
new MyClass().Should().Match<MyClass>(e => e.MyProperty == 42);
```

**Comment by jnyrup** [comment]
@georgiosd Does this answer your question?

**Comment by georgiosd** [comment]
Hey, sorry that I didn't follow up. It does, but explicitly defining the type in `Match<T>` seems very tedious. 

Wondering how this hasn't come up before! Maybe I'm somehow an edge case :)

**Comment by jnyrup** [comment]
It boils down to the difficulty of designing a library based on extension methods.
That means we rely deeply on overload resolution.
Currently `Match()` maps to `Match<TSubject>()` where `TSubject` is a generic type parameter of `ReferenceTypeAssertions<TSubject>`.
For your custom class, `Should()` creates an instance of `ObjectAssertions` which inherits from `ReferenceTypeAssertions<object>`.
This means that `o.Should().Match()` maps to `o.Should().Match<object>()`.

If we add `GenericObjectAssertions<T> : ReferenceTypeAssertions<T>` and the corresponding `public static GenericObjectAssertions<T> Should<T>(this T actualValue)`, many types will resolve `Should()` to that instead of the desired one.
E.g. `IEnumerable`, which should use `NonGenericCollectionAssertions` is now picked up by `GenericObjectAssertions`.

**Comment by georgiosd** [comment]
Ah, that makes sense. Because there's no constraint that could enforce non-IEnumerables.

What if there was a method in `ObjectAssertions` that would specialize it? Like `As<T>` but only for this purpose?

**Comment by jnyrup** [comment]
A specialization as in the existing `Match<T>()` assertion?

If you mean something like  
```c#
o.Should().Specialize<Myclass>().Should().Match()
``` 
then I prefer the current (and shorter)

```c#
o.Should().Should().Match<MyClass>()
```

**Comment by georgiosd** [comment]
One that will trickle down the chain:

`Should().As<T>().Match(...).Match(...)`

or even

`Should<T>().Match(...).Match(...)`

**Comment by jnyrup** [comment]
One could maybe do something like

```c#
class GenericObjectAssertions<T> : ReferenceTypeAssertions<T, GenericObjectAssertions<T>>
{
    public GenericObjectAssertions(object value)
    {
        if (value is null)
        {
            Subject = default(T);
        }
        else if (value is T castedValue)
        {
            Subject = castedValue;
        }
        else
        {
            throw new InvalidCastException();
        }
    }
}

GenericObjectAssertions<T> Should<T>(this object actualValue)
{
    return new GenericObjectAssertions<T>(actualValue);
}
```

That would be beneficial for 
```c#
ReferenceTypeAssertions.Match(Expression<Func<TSubject, bool>> predicate, ...)
```
as the right `TSubject` would be picked up

but worse(?) for 
```c#
ReferenceTypeAssertions.BeSameAs(TSubject expected, ...)
ReferenceTypeAssertions.NotBeSameAs(TSubject unexpected, ...)
```
as those are then "restricted" to `TSubject` instead of `object`.

**Comment by georgiosd** [comment]
Yes, I see.

But if it's a secondary method like `As<T>` then it shouldn't be a problem right?

**Comment by jnyrup** [comment]
Another workaround is to define a your own assertion class that derives from `ReferenceTypeAssertions`.
This of course gets tedious if you rely on `Match` for many different types.

```cs
private class MyClassAssertions : ReferenceTypeAssertions<MyClass, MyClassAssertions>
{
    public MyClassAssertions(MyClass subject)
        : base(subject)
    {
    }

    protected override string Identifier => nameof(MyClass);
}

public static Extensions
{
    public static MyClassAssertions Should(this MyClass myClass) =>
        new MyClassAssertions(myClass);
}
```

**Comment by CaringDev** [comment]
This would also be beneficial for custom generic extensions like: `foo.Should().HaveValidationErrorFor(it => it.Bar);`

Maybe something to consider for v7?

**Reply by dennisdoomen** [reply]
> Maybe something to consider for v7?

This would required a complete redesign of the API, so that's not going to happen.
